mlc_llm/transform/fuse_split_rotary_embedding.py

```diff
@@ -10,234 +10,295 @@
-                    A[v_ax0, v_ax1, v_ax2],
-                    A[v_ax0, v_ax1, v_ax2 + T.int64(hidden_size)],
-                    A[v_ax0, v_ax1, v_ax2 + T.int64(hidden_size * 2)],
-                )
-                T.writes(
-                    T_split[v_ax0, v_ax1, v_ax2],
-                    T_split_1[v_ax0, v_ax1, v_ax2],
-                    T_split_2[v_ax0, v_ax1, v_ax2],
-                )
-                pos: T.float32 = T.Cast("float32", n - T.int64(1))
+
+        for iters in T.grid(batch_size, seq_len, num_query_heads + num_kv_heads, head_dim):
+            with T.block("RotaryQueryKeyAndSplit"):
+                batch_i, seq_i, head_num, head_i = T.axis.remap("SSSS", iters)
+                pos: T.float32 = T.Cast("float32", rotary_offset - T.int64(1))
                 inv_freq: T.float32 = T.float32(1) / T.pow(
-                    T.float32(position_embedding_base),
-                    T.Cast("float32", (v_ax2 * 2) % head_dim) / T.float32(head_dim),
+                    position_embedding_base,
+                    T.Cast("float32", (head_i * 2) % head_dim) / T.float32(head_dim),
                 )
                 freq: T.float32 = pos * inv_freq
                 cos_value: T.float16 = T.Cast("float16", T.cos(freq))
                 sin_value: T.float16 = T.Cast("float16", T.sin(freq))
-                T_split[v_ax0, v_ax1, v_ax2] = cos_value * A[
-                    v_ax0, v_ax1, v_ax2
-                ] + sin_value * T.Select(
-                    T.int64(head_dim // 2) <= v_ax2 % head_dim,
-                    A[v_ax0, v_ax1, v_ax2 - T.int64(head_dim // 2)],
-                    A[v_ax0, v_ax1, v_ax2 + T.int64(head_dim // 2)] * T.float16(-1),
-                )
-                T_split_1[v_ax0, v_ax1, v_ax2] = cos_value * A[
-                    v_ax0, v_ax1, v_ax2 + T.int64(hidden_size)
+
+                embedded_value = cos_value * Fused_QKV[
+                    batch_i, seq_i, head_num, head_i
                 ] + sin_value * T.Select(
-                    T.int64(head_dim // 2) <= v_ax2 % head_dim,
-                    A[v_ax0, v_ax1, v_ax2 + T.int64(hidden_size) - T.int64(head_dim // 2)],
-                    A[v_ax0, v_ax1, v_ax2 + T.int64(hidden_size) + T.int64(head_dim // 2)]
+                    head_i < T.int64(head_dim // 2),
+                    Fused_QKV[batch_i, seq_i, head_num, head_i + T.int64(head_dim // 2)]
                     * T.float16(-1),
+                    Fused_QKV[batch_i, seq_i, head_num, head_i - T.int64(head_dim // 2)],
                 )
-                T_split_2[v_ax0, v_ax1, v_ax2] = A[v_ax0, v_ax1, v_ax2 + T.int64(hidden_size * 2)]
-
-    return split_rotary
+                if head_num < num_query_heads:
+                    EmbeddedQuery[batch_i, seq_i, head_num, head_i] = embedded_value
+                else:
+                    EmbeddedKey[batch_i, seq_i, head_num - num_query_heads, head_i] = embedded_value
 
+        for iters in T.grid(*Value.shape):
```

Comment:
I think separating a block for `Value` would end up with two kernel launches for split_rotary, while the previous implementation (for the non-GQA version) is fully fused. Also cc @jinhongyii if he has any concern in merging the GQA variant into a single generator.

Reply:
this looks good from my pov :)

Reply:
I don't have any concerns other than this. No matter how we generate the split_rotary_functions, we need to ensure that split_rotary only generates one kernel in non-GQA version.